src/main/java/org/optaplanner/springboottaskassigning/solver/DefaultSolverManager.java

```diff
@@ -99,54 +98,72 @@ public void solve(Object problemId,
     @Override
     public void stopSolver(Object problemId) {
         logger.debug("Stopping solver of problemId ({}).", problemId);
-        if (!problemIdToSolverTaskMap.containsKey(problemId)) {
-            throw new IllegalArgumentException("Problem (" + problemId + ") was not submitted.");
+        SolverTask<Solution_> solverTask;
+        synchronized (this) {
+            if (!isProblemSubmitted(problemId)) {
+                throw new IllegalArgumentException("Problem (" + problemId + ") was not submitted.");
+            }
+            solverTask = problemIdToSolverTaskMap.get(problemId);
+        }
+        if (solverTask != null) {
+            solverTask.stopSolver();
         }
-        problemIdToSolverTaskMap.get(problemId).stopSolver();
     }
 
     @Override
-    public boolean problemSubmitted(Object problemId) {
+    public boolean isProblemSubmitted(Object problemId) {
         return problemIdToSolverTaskMap.containsKey(problemId);
     }
 
     @Override
     public Solution_ getBestSolution(Object problemId) {
         logger.debug("Getting best solution of problemId ({}).", problemId);
-        if (!problemIdToSolverTaskMap.containsKey(problemId)) {
-            logger.error("Problem (" + problemId + ") was not submitted.");
+        SolverTask<Solution_> solverTask;
+        synchronized (this) {
```

Comment:
Now I realized there is much easier solution:
```
solverTask = problemIdToSolverTaskMap.get(problemId);
if (solverTask == null) {
    logger.error("Problem (" + problemId + ") was not submitted.");
}
return solverTask == null ? null : solverTask.getBestSolution();
```

Reply:
The race condition still exists though, now that I think about it, it makes more sense to throw an exception when asking for a non-existing solverTask.
The reason is there are two ways for this method to return null:
1. solverTask is null: problem was not submitted
2. bestSolution is null: problem is submitted but has not started solving yet
Currently there's no way the user will know programmatically (without reading the error logs) the exact reason. I'll add it for discussion and address it in a separate PR.

Reply:
Truth, it would be good to distinguish these two states by throwing the exception for a problem not being submitted and by using Optional :) if the best solution is not yet available.